File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations

def eratos(isPrime, integer):
    isPrime[0] = False
    isPrime[1] = False
    
    m = int(integer**0.5)

    for i in range(2, m + 1):
        if isPrime[i] == True:
            for j in range(i+i, integer+1, i):
                isPrime[j] = False

def solution(numbers):
    numbers = list(numbers) #문자열로 들어온 숫자를 리스트로 변환
    integers = []

    #길이만큼 순열로 뽑아야 하기 때문에 len(numbers)+1
    for i in range(1, len(numbers)+1):
        #조합으로 숫자 만들기
        c_list = list(permutations(numbers, i))
        for c in c_list:
            integers.append(int(''.join(list(c))))
    
    #중복 숫자 제거하기
    integers = sorted(list(set(integers)))
    maximum_number = integers[-1]

    #소수인지 아닌지 체크하는 배열
    isPrime = [True for _ in range(maximum_number+1)] 
    eratos(isPrime, maximum_number)

    answer = 0
    for integer in integers:
        if isPrime[integer] == True: answer+=1
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <set>
#include <cmath>

using namespace std;

set<int> s;
vector<bool> visited;

bool is_prime(int n){
    if (n <= 1) return false;
    for (int i=2; i<=sqrt(n); i++) {
        if(n % i == 0) return false;
    }
    return true;
}

void dfs(string numbers, string current) {
    if (!current.empty()){
        s.insert(stoi(current));
    }
    
    for (int i=0; i<numbers.size(); i++) {
        if (!visited[i]) {
            visited[i] = true;
            dfs(numbers, current+numbers[i]);
            visited[i] = false;
        }
    }
}

int solution(string numbers) {
    visited = vector<bool>(numbers.size(), false);
    
    dfs(numbers, "");
    
    int answer = 0;
    for (auto num: s) {
        if (is_prime(num)) answer++;
    }
    
    return answer;
}